### PR TITLE
Fix losing Exit event on scroll

### PR DESCRIPTION
**Pull request by igordmn**
Fixes https://github.com/JetBrains/compose-jb/issues/1324#issuecomment-980509700

Compose doesn't work well if we send an event with different coordinates without sending Move event before it:
```
Column {
  Box(size=10)
  Box(size=10)
}
```
If we send these events:
Move(5,5) -> Scroll(5,15) -> Move(5,15) -> Move(5,15)

Then during the scroll event, HitPathTracker forgets about the first Box, and never send Exit to it. Instead it sends the Scroll event to it.

We should send events in this order:
Move(5,5) -> Move(5,15) -> Scroll(5,15) -> Move(5,15)

With synthetic events things more complicated, as we always send events with the same position. I suppose the proper fix should be in Compose core itself, but it would a very huge fix.

Reproducer:
```
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.size
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.ImageComposeScene
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.PointerEventType
import androidx.compose.ui.input.pointer.onPointerEvent
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.dp
import androidx.compose.ui.use
import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@OptIn(ExperimentalComposeUiApi::class)
@RunWith(JUnit4::class)
class MouseHoverFilterTest {
    // bug: https://github.com/JetBrains/compose-jb/issues/1324#issuecomment-980509700
    @Test
    fun `move from one component to another, between another non-move event`() = ImageComposeScene(
        width = 100,
        height = 100,
        density = Density(1f)
    ).use { scene ->
        var enterCount1 = 0
        var exitCount1 = 0
        var enterCount2 = 0
        var exitCount2 = 0

        scene.setContent {
            Column {
                Box(
                    modifier = Modifier
                        .pointerMove(
                            onEnter = { enterCount1++ },
                            onExit = { exitCount1++ }
                        )
                        .size(10.dp, 10.dp)
                )
                Box(
                    modifier = Modifier
                        .pointerMove(
                            onEnter = { enterCount2++ },
                            onExit = { exitCount2++ }
                        )
                        .size(10.dp, 10.dp)
                )
            }
        }

        scene.sendPointerEvent(PointerEventType.Enter, Offset(5f, 5f))
        assertThat(enterCount1).isEqualTo(1)
        assertThat(exitCount1).isEqualTo(0)
        assertThat(enterCount2).isEqualTo(0)
        assertThat(exitCount2).isEqualTo(0)

        // Compose doesn't work well if we send an event with different type and different coordinates, without sending move event before it
        scene.sendPointerEvent(PointerEventType.Scroll, Offset(5f, 15f))
        scene.sendPointerEvent(PointerEventType.Move, Offset(5f, 15f))
        assertThat(enterCount1).isEqualTo(1)
        assertThat(exitCount1).isEqualTo(1)
        assertThat(enterCount2).isEqualTo(1)
        assertThat(exitCount2).isEqualTo(0)
    }
}
```

Another issue:
https://github.com/JetBrains/compose-jb/issues/1480

### CODE REVIEW

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/mouse/MouseScrollFilter.skiko.kt [hunk 1]
```diff
@@ -133,17 +133,17 @@ private suspend fun PointerInputScope.awaitScrollEvent() = awaitPointerEventScop
...
+    val scrollAmount = this?.scrollAmount ?: 1
     return MouseScrollEvent(
         delta = if (scrollType == MouseWheelEvent.WHEEL_BLOCK_SCROLL) {
             MouseScrollUnit.Page(value * scrollAmount)
         } else {
             MouseScrollUnit.Line(value * scrollAmount)
         },
-
-        // There are no other way to detect horizontal scrolling in AWT
-        orientation = if (isShiftDown || scrollDelta.x != 0f) {
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/mouse/MouseScrollFilter.skiko.kt:146]: We check isShiftDown on ComposeLayer level, we don't need it here

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/mouse/MouseScrollFilter.skiko.kt [hunk 2]
```diff
@@ -114,7 +114,7 @@ fun Modifier.mouseScrollFilter(
     //  the more proper behaviour would be to batch multiple scroll events into the single one
     while (true) {
         val event = awaitScrollEvent()
-        val mouseEvent = (event.mouseEvent as? MouseWheelEvent) ?: continue
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/input/mouse/MouseScrollFilter.skiko.kt:117]: We don't send a native event via ImageComposeScene, so we should provide default values for scrolling settings

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeLayer.desktop.kt [hunk 3]
```diff
@@ -205,21 +205,12 @@ internal class ComposeLayer {
         }
 
         override fun scheduleSyntheticMoveEvent() {
+            needSendSyntheticMove = true
```
Comment (olonho) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposeLayer.desktop.kt:209]: Why this change?

**Comment by igordmn** [comment]
Also, even without our synthetic events, AWT/platform can miss sending Move event:

```
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import java.awt.Dimension
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionListener
import javax.swing.JFrame

fun main() {
    runBlocking(Dispatchers.Main) {
        val frame = JFrame()
        frame.size = Dimension(300, 300)
        frame.addMouseWheelListener { e ->
            println("scroll ${e.x} ${e.y}")
            Thread.sleep(50)
        }
        frame.addMouseMotionListener(object : MouseMotionListener {
            override fun mouseDragged(e: MouseEvent?) {
            }

            override fun mouseMoved(e: MouseEvent) {
                println("move ${e.x} ${e.y}")
                Thread.sleep(50)
            }
        })
        frame.isVisible = true
    }
}
```

Result:
```
move 282 164
move 232 180
move 128 223
scroll 216 190  // Compose will mess its state here, if we don't send `move 216 190`
scroll 166 213
move 99 239
scroll 104 234
scroll 97 239
scroll 97 239
scroll 97 239
```

In that case, we also need to send synthetic Move event before any non-move event. That is because the bug is still exists [here](https://github.com/JetBrains/compose-jb/issues/1324#issuecomment-981148420).

**Review by olonho — APPROVED** [review]
